src/core/net/dns_headers.hpp

```diff
@@ -374,6 +374,182 @@ class Header : public Clearable<Header>
+     * E.g., "example.com", "example.com." (same as previous one), local.", "default.service.arpa", "." or "" (root).
+     *
+     * This method validates that the @p aName is a valid domain name format, i.e. no empty labels, and labels are
+     * `kMaxLabelLength` (63) characters or less, and the domain name is `kMaxLength` (255) characters or less.
+     *
+     * @param[in]  aName              A domain name string. Can be nullptr (then treated as "." or root).
+     * @param[in]  aMessage           The message to append to.
+     *
+     * @retval OT_ERROR_NONE          Successfully encoded and appended the domain name to @p aMessage.
+     * @retval OT_ERROR_INVALID_ARGS  Domain name @p aName is not valid.
+     * @retval OT_ERROR_NO_BUFS       Insufficient available buffers to grow the message.
+     *
+     */
+    static otError AppendName(const char *aName, Message &aMessage);
+
+    /**
+     * This static method parses and skips over a full domain name in a message.
+     *
+     * @param[in]    aMessage         The message to parse the domain name from.
+     * @param[inout] aOffset          On input the offset in @p aMessage pointing to the start of the name field.
+     *                                On exit (when parsed successfully), @p aOffset is updated to point to the byte
+     *                                after the end of domain name field.
+     *
+     * @retval OT_ERROR_NONE          Successfully parsed and skipped over domain name, @p Offset is updated.
+     * @retval OT_ERROR_PARSE         Domain name could not be parsed (invalid format).
+     *
+     */
+    static otError ParseName(const Message &aMessage, uint16_t &aOffset);
+
+    /**
+     * This static method reads a name label from a given message.
+     *
+     * This method can be used to read labels one by one in a domain name. After a successful label read, @p aOffset is
+     * updated to point to the start of the next label. When we reach the end of the name, OT_ERROR_NOT_FOUND is
+     * returned. This method handles compressed domain names which use pointer labels. So as the labels in a name are
+     * read, the @p aOffset may jump back in the message and at the end the @p aOffset does not necessarily point to
+     * the end the original name field.
+     *
+     * @param[in]    aMessage         The message to read the label from.
+     * @param[inout] aOffset          On input, the offset in @p aMessage pointing to the start of the label to read.
+     *                                On exit, when successfully read, @p aOffset is updated to point to the start of
+     *                                the next label.
+     * @param[in]    aHeaderOffset    The offset in @p aMessage to the start of the DNS header.
+     * @param[out]   aLabelBuffer     A pointer to a char array to output the read label as a null-terminated C string.
```

Comment:
What about also outputting the label length? It could be useful if we want to join multiple labels in a loop.

Reply:
We can return the label length in `aLabelBufferSize` (probably rename the variable). I was considering it but was not sure if would be immediately useful. Will add it later in the next push (I think it would not add much complexity/overhead and it is an added functionality).

----

Later, we can consider adding method `ReadName()` to read and create a dot separated domain name string. We should be careful of cases where a (first) label can have dot `'.'` char in it (e.g., Service Instance Name). The approach seems to be to escape the dot char with `\` (and then if there are `\` chars in the label they become `\\`).

Reply:
Added the update on `aLabelLength` from `ReadLabel()` in the new pushed commit.

Reply:
> Later, we can consider adding method ReadName() to read and create a dot separated domain name string.

This is what I am planning to add as this is immediately useful on the server side, good to have it if you can add it in this PR.

> We should be careful of cases where a (first) label can have dot '.' char in it (e.g., Service Instance Name). 

Sorry I didn't get the point. Since the name is already encoded as a sequences of labels each with a preceding length field, it looks to me that we only need to append a `.` at end of each label to create the full name.

Not sure if the dots in the instance or host name worth a special handling. Refusing such names at the client side will work for me, although other DNS implementations may allow this. I believe `allowing dot in service instance name` is a mistake :(

Reply:
I can add `ReadName()` from this PR (will add from next next push). 😃 

---- 

From my understanding/reading, the "Instance Service Name" label seems to be treated differently from other labels. It is intended to be a human readable name so can have more characters than allowed by other labels (details in [RFC-6763-4.1.1](https://tools.ietf.org/html/rfc6763#section-4.1.1) and more in appendix). There is also a section suggesting the escaping model for the Instance Name [RFC6764-4.3](https://tools.ietf.org/html/rfc6763#section-4.3). I just mentioned this for us to be aware of it. 

For now we can have `ReadName()` assume there is no dot in the any of labels. We will see if we need to handle the Instance Service Name differently later. We can always read the Instance Service Name label separately using `ReadLabel()` which should handle it correctly.

Reply:
Added `RaadName()` in new pushed commit.